### DISCUSSION TITLE
Using CAN examples?

### DISCUSSION THREAD

**Discussion by jlmxyz** [CAP1Sup/Intellistep#19] [General]
I wonder if you have any documentation how to use CAN interface ? I'm building a CNC with dual closed loop controled motors for Y and Z axis, right now using STEP/DIR interface, but I think that CAN could be better....

**Comment by CAP1Sup** [comment]
> I wonder if you have any documentation how to use CAN interface ? I'm building a CNC with dual closed loop controled motors for Y and Z axis, right now using STEP/DIR interface, but I think that CAN could be better....

As of right now, you cannot use CAN at all due to a pin allocation issue. I will be fixing CAN functionality soon. There is currently no support for position commands, but I suppose that the commands could be added with only a little work. I'll add it to the long term feature plan. As of right now, I'm still working on making sure that the motor steps correctly and hopefully getting it a little quieter

**Reply by jlmxyz** [reply]
Ok, was just wondering because it´s running into my mind since a long time.... I wonder if it´s possible to do a "distributed" control system....
I was thinking of re-designing MKSs57B board or BTTservo57B board and add the following : 

1. thermal sensor support (ADC)
2. 2 stop switches support (GPIO)
3. one level probe support (either protected GPIO or HAL sensor)
3. RPM sensor (still don´t know how to make it... maybe will need an ADC and some linear frequency to voltage convertor)
4. 1 protected GPIO to drive a relay

the idea behind is to use CAN to have the motor controlled, and have the closed loop control reports some status like 

1. motor jamed
2. start position reached
3. max position reached
4. thermal sensor temperature (if the motor is used to drive PLA feeder, for a 3d printer for example, this could help shorten cables)
5. rpm if the motor is used close to the CNC main motor

I know marlin don´t support CAN and is not designed for distributed controls boards... I wonder if there is any software that do it and on what platform they are running.... 
one "idea" is to have 
G0 [E<pos>] [F<rate>] [X<pos>] [Y<pos>] [Z<pos>]
broadcasted on CAN, then have the X assigned motors perform the move job.... and reply M105 [T<index>] commands, M119....
and also report absolute position (allowing same axis assigned motors to sync position...)

**Reply by TazerReloaded** [reply]
Klipper supports multi-MCU printers, it does the clock synchronisation and command distribution automatically. It does not currently support CAN or the BTT boards as far as I know, but there were already discussions on adding it: https://github.com/KevinOConnor/klipper/issues/3103
It's a quite complex topic, but if someone were to make it production ready, I'm sure it would be merged into the master branch.

**Reply by CAP1Sup** [reply]
If you wanted to do that, you're welcome to. Intellistep could be probably be easily configured to support the new board as well while still providing the S42BV2 functionality. I'm happy to get the CAN working, it's something that I'd really like to enjoy. Unfortunately for now, I'm going to focus on just the motor and getting the basic functionality working. After that, I want to add the dynamic current control. Then I'll add all of the CAN functionality. The project is useless if it can't fulfill its main function, and the dynamic current would be very nice to have. I'll let you know once I complete the CAN interface for it. In the meantime, if you wanted to get things started on the Klipper / Marlin side, then we could move straight to testing instead of having to add the functionality to Klipper / Marlin first. Let me know how it goes if you do decide to work on adding the functionality

**Reply by jlmxyz** [reply]
didn´t knew about klipper.... and seems that klipper support CAN bus also!!  https://www.klipper3d.org/CANBUS_protocol.html so if you´re about to rewrite the CAN layer, better make it compatible with klipper low layer protocol

using an octoprint + klipper plugin + usb2CAN + redesigned s42b closed loops boards  begins to look like a really nice idea!!!

**Reply by jlmxyz** [reply]
I cloned the repository, opened it in vscode+platformio plugin, and begun to look at it for the CAN, right now just wanting to give a try without modifications, build went fine, but wonder why there is no main()?

**Reply by jlmxyz** [reply]
btw is the current state "likely usable"? I don´t speak about accuracy or noise, but just moving and reporting? I think I´ll first use serial line since I´ve an usb to serial to give a try to klipper and add the klipper protocol to it...

**Reply by CAP1Sup** [reply]
Unfortunately not yet. I've been having some issues with reverse direction on the motor, so it's been taking me a longer time. If you wanted to get started on Klipper, go right ahead and start with your own fork. I'll be happy to help and merge it once the feature is working. If you did start on it, would you mind making a separate branch for it? That way, when your changes are merged into Intellistep, we can continue to work on it without affecting the working build. Also, I'd like to see it set up Klipper as an added feature, so building the support into the existing parser and adding compiler directives (like #ifdef) would be ideal as it would keep the backwards compatibility. I'm hoping to get the code finished up soon, but I've been busy with work lately and haven't had time at home to work on the actual stepping procedure. It should be finished within the month. I'd love to have support on the boards for both Marlin and Klipper, as well as features such as position correction.

**Reply by CAP1Sup** [reply]
The only concern that I have is that the CAN voltage might not be 5v. You'd have to check on that, otherwise you could accidentally cause the magic smoke to escape.

**Comment by jlmxyz** [comment]
hummmm according to my investigations,
3.3V is provided by U8 which is a 1117 3.3 regulated power (either 1A or
800mA depending of manufacturer)
the input pin of U8 is connected to a chip marked "BNOR 11" pin3or4 through
a 15kohm restistor, there is a huge self right beside this chip and pin2or5
is connected to VMOT through a huge diode....
I made some mesurements, with +5V of DIR/STEP interface unpluged (just let
the VMOT, GND, EN, DIR, STP), the board still works and VCC pin of
PCA82C251 is 5V
so it seems that if BTT software engineer are bad, electrical ones did the
work by far better than the MKS servo engineers....
1) they use VMOT to create internal power, which is by far better
2) VMOT is droped to 5V using the unknown chip BNOR11,
3) 5V is then regulated to 3.3V using the 1117 3.3 regulated power...
too bad they didn´t used pins that could do analog sampling for the
switches, else I could have replaced the switches by a connector and used
them for end stops and thermal... without having to do huge redesing...
so at least the CAN chip is correctly powered.

I´ll start work on klipper, I think I´ll add a menu in the display to
activate or not the support of the protocol, once I understood how they do
it...

don´t like too much #ifdef... it makes the code unreadble... better use
compiler optimizations to whipe the dead code using
#define KLIPPER_SUPPORT 0

if(KLIPPER_SUPPORT)
{
....
}

this way you can still indent the code and allow folding


Le sam. 15 mai 2021 à 20:11, Christian Piper ***@***.***> a
écrit :

**Comment by jlmxyz** [comment]
I've some question because I started schematics for the new board.
the motor are driven by analog write on PB4 and PB5, but I don't find where
theses pins are setup and if they are connected to PWM's timer1 of
stm32f103R8....
where are the pins configured?
thanks and regards

**Reply by CAP1Sup** [reply]
Here they are: 
https://github.com/CAP1Sup/Intellistep/blob/43a391a9c9379926e8b7b8ce2fdb2c2c8330125c/src/hardware/motor.cpp#L39-L44

I figured that I would put the code in the constructor of the motor class as it is related to the motor's start up and necessary initialization

**Reply by CAP1Sup** [reply]
As for the timer, they are connected on channel 1 and 2 of timer 3 of the STM32. That's a hardware linkage that is handled automatically by the Arduino library. Keep in mind that you will have to figure out the correct CAN bus pins, as the ones that I had listed are wrong (I think, check with BTT's code)

**Reply by jlmxyz** [reply]
ideed, yours are 
```
// CAN bus pins
#define CAN_IN_PIN   PA_11
#define CAN_OUT_PIN  PA_12
```

and code in legacy has 
```
/**TIM1 GPIO Configuration  
  PA12   ------> TIM1_ETR 
  */
  GPIO_InitStruct.Pin = LL_GPIO_PIN_12;
  GPIO_InitStruct.Mode = LL_GPIO_MODE_ALTERNATE;
  GPIO_InitStruct.Speed = LL_GPIO_SPEED_FREQ_HIGH;
  GPIO_InitStruct.OutputType = LL_GPIO_OUTPUT_PUSHPULL;
  GPIO_InitStruct.Pull = LL_GPIO_PULL_NO;
  GPIO_InitStruct.Alternate = LL_GPIO_AF_2;
  LL_GPIO_Init(GPIOA, &GPIO_InitStruct);
```

-_-

**Reply by CAP1Sup** [reply]
Alright, my CAN pins were correct. The issue is that the CAN setup code didn't use the pins in the config file.

**Reply by jlmxyz** [reply]
according to datasheet, CAN ports can be remapped to 
RX : PA11, PB8, PD0
TX : PA12, PB9, PD1
unfortunatly none are described in legacy code.... so I guess I´ll have to test electrically...  :(

**Reply by CAP1Sup** [reply]
I'm fairly confident that the code uses the PA11 and 12 pins. They just used timers to trigger the communication lines

**Reply by jlmxyz** [reply]
ok PA11&PA12... I´ve 0 ohms between those pins and input pins of can tranceiver.... how awful code it is...

**Reply by CAP1Sup** [reply]
Exactly, one of the original reasons I decided to start this project

**Reply by jlmxyz** [reply]
I´m a bit concerned because in main.h and main.c of legacy, I didn´t found what you put into config.h

here is a begin of reverse eng schematic

**Reply by CAP1Sup** [reply]
I relied on the original BTT code for all of the config.h documentation. A schematic would be nice though

**Reply by jlmxyz** [reply]
![image](https://user-images.githubusercontent.com/40383940/118416018-de4b3400-b6ad-11eb-87d6-88fda8b6d124.png)

**Reply by jlmxyz** [reply]
this is according to main.h and main.c.... but I´m annoyed by the PA11 which is used both for button and CAN_RX

**Reply by CAP1Sup** [reply]
Did you double check PA11 and 12? I might have been wrong from the code analysis.

**Reply by jlmxyz** [reply]
![image](https://user-images.githubusercontent.com/40383940/118772853-80247980-b884-11eb-8780-46857234f35f.png)
ncurrent reverse state directly from electrical checking, need help to know what K1, K2, K3 switches are for and I'm a bit puzzled by the 1&2 dip switches that are tied to VSS*

**Reply by jlmxyz** [reply]
Ok here is almost complete and Cpu Re-checked schematic, miss the Vmot->5V part

**Reply by jlmxyz** [reply]
![BTT_S57B_v2 0_reverseIng 0 1 0](https://user-images.githubusercontent.com/40383940/119202204-956df380-ba90-11eb-9760-68fb4cb807aa.png)

**Reply by jlmxyz** [reply]
what is really interesting, is that we can un-solder the dip switches and make this board a complete Klipper controler board...
PA15 used for calibsw have analog => can be used to attach a board for termal sensor
PB11 & PB10 are attached to TIM2_CH4 and TIM2_CH3.... maybe they can be used for PWM driving fans??? (EDIT : yes they can be used to drive FAN!!!! it will require to add control board but this can be done as a module attached near to the FAN)
PB3 can be used to attach a board and start/end switches.....

this is really interesting with this board we can have a klipper subsystem that can act as
1) motor driver
2) thermal sensor
3) one fan control and 2 end/start switches detector
or
3bis) two fan control and start switch
or
3ter) two fan, 2 switches in parallel, start or end is found according to motor direction... if one switch triggers, if we were moving toward start it's a "homing reached event" if we were moving toward end, it's "exceed limit event"

and more, s42b has no can, but this board as UART so can act as a proxy.... ok, I begin to love this idea!!!

**Reply by jlmxyz** [reply]
Ok, I started the design of the add-on board, and I'm a bit sad... in fact PA15 hasn't ADC input possibility...
so, I'm a bit annoyed....
it is possible but not using the dipswitches but the press button, must look if I find some connector that can go in place...
what can be done if using the v2.0 board as is 

1. unsolder dipswitches, press buttons and C19 and r20
2. solder on fingerprints some connector
3. have the add-on board handle the press button with following remappings : 
        1. PB0 from Mode-K3 to Thermal sensor
        2. PB1 from Valid-k4 to TIM3-Ch4 or TIM1-CH3 (FAN)
        3. PA3 from Next-k2 to TIM2_CH4 (FAN)
        4. PA15 from CALIB_SWITCH to END-START switch (with moving direction determining the end or start condition)

the issue with that is that I wanted to keep the possibility to just unplug the addon board and go back to previous behavior with for example a menu on the oled (as well as activating the klipper mode into the menu....) but with this hardware modifications the add on board will be required to operate the oled, thus there is no "software switch".....

another thing, I begun to look at klipper code, they don't use arduino libs but regular stm32 libs, arduino libs don't support CAN for stm32 if I'm correct, this means that we will have to include both libraries and I'm not sure it won't over consume flash... 
is arduino libs mandatory? 
didn't looked at your code, to know if it's an easy switch from one solution to another....

**Reply by CAP1Sup** [reply]
If you wanted to include the STM code, then we could covert/adapt it. The CAN library is already included in the betterReverse branch, awaiting debugging. Most of the stm32f1 libraries are included with the Arduino, so I'll have to take a look at the code you're referring to. Good luck on the addition board, I can add an optional setting to enable or disable the related code

**Reply by jlmxyz** [reply]
klipper CAN code : https://github.com/KevinOConnor/klipper/blob/master/src/stm32/can.c
anyway, if want to make a can bus, the R19 resistor must be removed, appart on one device, since the CAN require a single 120Ohm resistor per bus
I just begins to look at klipper code, in particular for messaging, I guess the easiest way will be to have the klipper code stored and compile only the required parts,
I don´t know if it might helps, but klipper can perform some kind of buffering, maybe this could help for acceleration computing and then noise reduction

**Reply by bilalibrir** [reply]
[TTB_S57B_REV1.pdf](https://github.com/CAP1Sup/Intellistep/files/6712308/TTB_S57B_REV1.pdf)
hello everyone I'm happy to share my version of the BTT S57B stepper controller.
there is still a lot of work to do especially because I don't own any module, and everything is done from some pictures and from the code.
I'm trying to make a design that uses the blue pill board.

**Comment by jlmxyz** [comment]
hummm ok you use the arduino lib.... I'm not used to it, I done freertos or
the like dev, never arduino... so I don't know how the pins are
configured...
one other thing, I've also s42Bv1 board, the MCU is a STM32F030C8....
didn't looked into datasheet yet, but do you think your Firmware will be
compatible? because they seems broken as well (sensor seems same and
overall design seems the same only can isn't present on this board)

Le dim. 16 mai 2021 à 20:28, Christian Piper ***@***.***> a
écrit :

**Reply by CAP1Sup** [reply]
I'm hoping it will be compatible. The pin configs will have to be moved to their own board files. Other than that, there isn't a lot of changes that need to be done

**Comment by IhorNehrutsa** [comment]
Here CAN is remapped to PB8, PB9, but the CAN driver is connected to PA11 and PA12 on the schematic.
https://github.com/CAP1Sup/Intellistep/blob/49a0e18ae28020164c2984caa41fc91d0b539355/src/lib/CAN/CAN.cpp#L17-L27
And I find commented CANInit() and can't find another CAN code 
https://github.com/CAP1Sup/Intellistep/blob/49a0e18ae28020164c2984caa41fc91d0b539355/src/hardware/canMessaging.cpp#L16-L24
Please clarify for me about the CAN.

**Reply by CAP1Sup** [reply]
I disabled CAN in the master branch because I didn't know the correct pins and couldn't find a library at the time. The CAN implementation is fixed in the betterReverse branch, which will eventually become the dev branch. I still need to do testing, but all of the code should work theoretically. I'll do testing and debugging on it later, but right now my main focus is getting the motor to step correctly.
